File: MemeShareJetpack/app/src/main/java/com/example/memesharejetpack/view/AnasayfaScreen.kt
package com.example.memesharejetpack.view

import android.widget.Toast
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.example.memesharejetpack.viewmodel.OdaViewModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AnasayfaScreen(
    userId: Int,
    onOpenRoom: (roomId: Int, userId: Int) -> Unit,
    viewModel: OdaViewModel = viewModel()
) {
    val context = LocalContext.current

    // ViewModel verileri
    val rooms by viewModel.joinedRooms.observeAsState(emptyList())
    val createResult by viewModel.odaOlusturmaSonucu.observeAsState()
    val joinResult by viewModel.joinResult.observeAsState()

    // Dialog state
    var showJoinDialog by remember { mutableStateOf(false) }
    var roomCodeField by remember { mutableStateOf(TextFieldValue("")) }

    // İlk açılışta odaları çek
    LaunchedEffect(userId) {
        viewModel.fetchJoinedRooms(userId)
    }

    // Oda oluşturma sonucu side-effect
    LaunchedEffect(createResult) {
        createResult?.let { res ->
            if (res.success) {
                Toast.makeText(context, "Oda oluşturuldu: ${res.roomCode}", Toast.LENGTH_SHORT).show()
                // Listeyi yenile
                viewModel.fetchJoinedRooms(userId)
            } else if (res.message.isNotBlank()) {
                Toast.makeText(context, "Hata: ${res.message}", Toast.LENGTH_SHORT).show()
            }
        }
    }

    // Odaya katılma sonucu side-effect
    LaunchedEffect(joinResult) {
        joinResult?.let { res ->
            if (res.success) {
                Toast.makeText(context, "Odaya katıldınız", Toast.LENGTH_SHORT).show()
                viewModel.fetchJoinedRooms(userId)
            } else if (res.message.isNotBlank()) {
                Toast.makeText(context, res.message, Toast.LENGTH_SHORT).show()
            }
        }
    }

    Scaffold(
        topBar = {
            TopAppBar(title = { Text("Odalarım") })
        },
        floatingActionButton = {
            ExtendedFloatingActionButton(
                onClick = { viewModel.createRoom(userId) },
                content = { Text("Oda Oluştur") }
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .padding(16.dp)
        ) {
            Button(
                onClick = { showJoinDialog = true },
                modifier = Modifier.fillMaxWidth()
            ) {
                Text("Odaya Katıl (Kod ile)")
            }

            Spacer(Modifier.height(12.dp))

            if (rooms.isEmpty()) {
                Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center
                ) { Text("Henüz oda yok") }
            } else {
                LazyColumn(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    items(rooms) { oda ->
                        RoomCard(
                            roomCode = oda.roomCode,
                            createdBy = oda.createdBy,
                            onClick = { onOpenRoom(oda.odaId, userId) }
                        )
                    }
                }
            }
        }
    }

    // Join dialog
    if (showJoinDialog) {
        AlertDialog(
            onDismissRequest = { showJoinDialog = false },
            title = { Text("Oda Katılım") },
            text = {
                OutlinedTextField(
                    value = roomCodeField,
                    onValueChange = { roomCodeField = it },
                    singleLine = true,
                    label = { Text("Oda Kodu") },
                    modifier = Modifier.fillMaxWidth()
                )
            },
            confirmButton = {
                TextButton(onClick = {
                    val code = roomCodeField.text.trim()
                    if (code.isNotEmpty()) {
                        viewModel.joinRoom(userId, code)
                        showJoinDialog = false
                        roomCodeField = TextFieldValue("")
                    }
                }) { Text("Katıl") }
            },
            dismissButton = {
                TextButton(onClick = { showJoinDialog = false }) { Text("İptal") }
            }
        )
    }
}

@Composable
private fun RoomCard(
    roomCode: String,
    createdBy: Int,
    onClick: () -> Unit
) {
    Card(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick),
        elevation = CardDefaults.cardElevation(defaultElevation = 4.dp)
    ) {
        Column(Modifier.padding(16.dp)) {
            Text(text = roomCode, style = MaterialTheme.typography.titleMedium)
            Spacer(Modifier.height(8.dp))
            Text(text = "Oluşturan: $createdBy", style = MaterialTheme.typography.bodyMedium)
        }
    }
}

File: meme_share_flutter/lib/view/anasayfa_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../components/snackbar.dart';
import '../viewmodel/oda_vm.dart';

class AnasayfaScreen extends StatefulWidget {
  final int userId;
  final void Function(int roomId, int userId) onOpenRoom;

  const AnasayfaScreen({
    super.key,
    required this.userId,
    required this.onOpenRoom,
  });

  @override
  State<AnasayfaScreen> createState() => _AnasayfaScreenState();
}

class _AnasayfaScreenState extends State<AnasayfaScreen> {
  @override
  void initState() {
    super.initState();
    // Compose LaunchedEffect(userId)
    WidgetsBinding.instance.addPostFrameCallback((_) {
      context.read<OdaVM>().fetchJoinedRooms(widget.userId);
    });
  }

  Future<void> _showJoinDialog(BuildContext context) async {
    final codeCtrl = TextEditingController();

    final ok = await showDialog<bool>(
      context: context,
      builder: (ctx) => AlertDialog(
        title: const Text('Oda Katılım'),
        content: TextField(
          controller: codeCtrl,
          decoration: const InputDecoration(labelText: 'Oda Kodu'),
        ),
        actions: [
          TextButton(onPressed: () => Navigator.pop(ctx, false), child: const Text('İptal')),
          TextButton(
            onPressed: () => Navigator.pop(ctx, true),
            child: const Text('Katıl'),
          ),
        ],
      ),
    );

    if (ok == true) {
      final code = codeCtrl.text.trim();
      if (code.isEmpty) return;

      final vm = context.read<OdaVM>();
      await vm.joinRoom(userId: widget.userId, roomCode: code);

      if (!mounted) return;

      if (vm.joinResult?.success == true) {
        showSnack(context, 'Odaya katıldınız');
        await vm.fetchJoinedRooms(widget.userId);
      } else {
        showSnack(context, vm.joinResult?.message ?? vm.error ?? 'Katılım başarısız');
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return Consumer<OdaVM>(
      builder: (context, vm, _) {
        final rooms = vm.joinedRooms;

        return Scaffold(
          appBar: AppBar(title: const Text('Odalarım')),
          floatingActionButton: FloatingActionButton.extended(
            onPressed: vm.isLoadingCreate
                ? null
                : () async {
              await vm.createRoom(userId: widget.userId);

              if (!mounted) return;

              if (vm.odaOlusturmaSonucu?.success == true) {
                showSnack(context, 'Oda oluşturuldu: ${vm.odaOlusturmaSonucu?.roomCode ?? ""}');
                await vm.fetchJoinedRooms(widget.userId);
              } else {
                showSnack(context, vm.odaOlusturmaSonucu?.message ?? vm.error ?? 'Oda oluşturma hatası');
              }
            },
            label: vm.isLoadingCreate ? const Text('...') : const Text('Oda Oluştur'),
          ),
          body: Padding(
            padding: const EdgeInsets.all(16),
            child: Column(
              children: [
                SizedBox(
                  width: double.infinity,
                  child: ElevatedButton(
                    onPressed: () => _showJoinDialog(context),
                    child: const Text('Odaya Katıl (Kod ile)'),
                  ),
                ),
                const SizedBox(height: 12),
                Expanded(
                  child: vm.isLoadingRooms
                      ? const Center(child: CircularProgressIndicator())
                      : rooms.isEmpty
                      ? const Center(child: Text('Henüz oda yok'))
                      : ListView.separated(
                    itemCount: rooms.length,
                    separatorBuilder: (_, __) => const SizedBox(height: 8),
                    itemBuilder: (ctx, i) {
                      final oda = rooms[i];
                      return Card(
                        child: ListTile(
                          title: Text(oda.roomCode),
                          subtitle: Text('Oluşturan: ${oda.createdBy}'),
                          onTap: () => widget.onOpenRoom(oda.odaId, widget.userId),
                        ),
                      );
                    },
                  ),
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}
